File: Plugins/Decoders/ProofPointDecoder.py
from Plugins import Plugin

import re
import urllib.parse
import html

class ProofPointDecoder(Plugin.Plugin):
  def __init__(self, url: str = None, name = "ProofPoint Decoder"):
    super().__init__(name)
    self.name = name
    self._url = url.strip() if url is not None else None
    self._url_re = r'https://urldefense.proofpoint.com/(v[0-9])/'
    self._urlv3_re = r'urldefense.com/(v3)/'
  
  def decodev1(self, rewrittenurl):
    match = re.search(r'u=(.+?)&k=', rewrittenurl)

    linksFoundList = []

    if match:
      urlencodedurl = match.group(1)
      htmlencodedurl = urllib.parse.unquote(urlencodedurl)
      url = html.unescape(htmlencodedurl)
      url = re.sub("http://", "", url)
      if url not in linksFoundList:
        linksFoundList.append(url)
    return linksFoundList

  def decodev2(self, rewrittenurl):
    match = re.search(r'u=(.+?)&[dc]=', rewrittenurl)

    linksFoundList = []

    if match:
      specialencodedurl = match.group(1)
      trans = str.maketrans('-_', '%/')
      urlencodedurl = specialencodedurl.translate(trans)
      htmlencodedurl = urllib.parse.unquote(urlencodedurl)
      url = html.unescape(htmlencodedurl)
      url = re.sub("http://", "", url)
      if url not in linksFoundList:
        linksFoundList.append(url)
    return linksFoundList

  def decodev3(self, rewrittenurl):
    match = re.search(r'v3/__(?P<url>.+?)__;', rewrittenurl)
    linksFoundList = []
    if match:
      url = match.group('url')
      if re.search(r'\*(\*.)?', url):
        url = re.sub('\*', '+', url)
        if url not in linksFoundList:
          linksFoundList.append(url)
    return linksFoundList

  def run(self):
    print("\n --------------------------------- ")
    print(" P R O O F P O I N T D E C O D E R ")
    print(" --------------------------------- ")
    if self._url == None:
      self._url = str(input("Enter URL to decode: ")).strip()

    print(self._url)
    match = re.search(self._url_re, self._url)
    matchv3 = re.search(self._urlv3_re, self._url)
    if match:
      if match.group(1) == 'v1':
        links = self.decodev1(self._url)
        for each in links:
          print('\nDecoded Link: %s' % each)
      elif match.group(1) == 'v2':
        links = self.decodev2(self._url)
        for each in links:
          print('\nDecoded Link: %s' % each)
    
    if matchv3 is not None:
      if matchv3.group(1) == 'v3':
        links = self.decodev3(self._url)
        for each in links:
          print('\nDecoded Link: %s' % each)
      else:
        print('No valid URL found in input: ', self._url)

File: src/plugins/decoders/proofpoint_decoder.rs
use super::Plugin;
use serde_json;
use regex::Regex;
use urlencoding::decode;

use crate::utils::get_input;

pub struct ProofPointDecode {
  url_re: Regex,
  url_re_v3: Regex
}

impl Plugin for ProofPointDecode {
  const NAME: &'static str = "ProofPoint Decoder";

  fn new() -> Self {
    Self {
      url_re: Regex::new(r"https://urldefense.proofpoint.com/(v[0-9])/").unwrap(),
      url_re_v3: Regex::new(r"urldefense.com/(v3)/").unwrap()
    }
  }

  fn perform_lookup(&self, value: &str) -> Result<serde_json::Value, String> {
    let match_result = self.url_re.find(value);
    let matchv3_result = self.url_re_v3.find(value);

    let mut decoded_links: Vec<String> = Vec::new();

    if let Some(m) = match_result {
      println!("{}", m.as_str());
      if String::from(m.as_str()).contains("v1") {
        let mut links = Self::decodev1(value);
        for item in &links {
          println!("\nDecoded Link: {}", item);
        }
        decoded_links.append(&mut links);
      } else if String::from(m.as_str()).contains("v2") {
        let mut links = Self::decodev2(value);
        for item in &links {
          println!("\nDecoded Link: {}", item);
        }
        decoded_links.append(&mut links);
      }
    }

    if let Some(m) = matchv3_result {
      if m.as_str() == "v3" {
        let mut links = Self::decodev3(value);
        for item in &links {
          println!("\nDecoded Link: {}", item);
        }
        decoded_links.append(&mut links);
      }
    }

    Ok(serde_json::json!({"status": "complete", "decoded_links": decoded_links}))
  }

  fn run(&self) {
    println!("{}", Self::NAME);
    
    let usr_in = get_input("Enter URL to decode: ");
    match self.perform_lookup(&usr_in.trim()) {
      Ok(result) => println!("\n{}", result),
      Err(error) => eprintln!("Error: {}", error)
    }
  }
}

impl ProofPointDecode {
  fn decodev1(rewrittenurl: &str) -> Vec<String> {
    let re: Regex = Regex::new(r"u=(.+?)&k=").unwrap();
    let mut links_found_list: Vec<String> = Vec::new();

    if let Some(caputres) = re.captures(rewrittenurl) {
      if let Some(encoded_url) = caputres.get(1) {
        let url_encoded_url = encoded_url.as_str();
        if let Ok(html_encoded_url) = decode(url_encoded_url) {
          let url = html_encoded_url.replace("http://", "");
          println!("{}", url);
          if !links_found_list.contains(&url) {
            links_found_list.push(url);
          }
        }
      }
    } else {
      println!("No captures for v1 found.")
    }

    links_found_list
  }

  fn decodev2(rewrittenurl: &str) -> Vec<String> {
    let re: Regex = Regex::new(r"u=(.+?)&[dc]=").unwrap();
    let mut links_found_list: Vec<String> = Vec::new();
  
    if let Some(captures) = re.captures(rewrittenurl) {
      if let Some(special_encoded_url) = captures.get(1) {
        let trans = special_encoded_url.as_str().replace("-", "%").replace("_", "/");
        if let Ok(url_encoded_url) = decode(&trans) {
          let html_encoded_url = url_encoded_url.replace("http://", "");
          if let Ok(url) = decode(&html_encoded_url) {
            println!("{}", url);
            if !links_found_list.contains(&url.to_string()) {
              links_found_list.push(url.to_string());
            }
          }
        }
      }
    } else { 
      println!("No captures for v2 found.")
    }
  
    links_found_list
  }

  fn decodev3(rewrittenurl: &str) -> Vec<String> {
    let re: Regex = Regex::new("v3/__(?P<url>.+?)__;").unwrap();
    let mut links_found_list: Vec<String> = Vec::new();

    if let Some(captures) = re.captures(rewrittenurl) {
      if let Some(url) = captures.name("url") {
        let mut url = url.as_str().to_string();
        if Regex::new(r"\*(\*.)?").unwrap().is_match(&url) {
          url = url.replace("*", "+");
          println!("{}", url);
          if !links_found_list.contains(&url) {
            links_found_list.push(url);
          }
        }
      }
    } else {
      println!("No captures for v3 found.")
    }
    
    links_found_list
  }
}